Extend to search and parse multiple message directories

euwest opened the pull request:
allows for passing `localeDirs`(an array of paths to directories containing locale directories with `po` files in them) to create message dictionaries by merging same-language `po`s.

example:

``` js

var options = {
    localeDirs: [
        'locale', 
        'node_modules/some-module/locale',
        'bower_components/some-component/locale'
    ]
};

gulp.task('translate', function(){
  return gulp.src('collected_static')
    .pipe(statici18n(options));
});
```

the first directory in `localeDirs` is treated as the canonical source, which means the languages the app is translated into are derived from the contents of this directory, and the messages from `.po` files in it are extended but not overwritten by proceeding directories.

euwest committed:
change tests to check for multiple message directories
- add nested-msgs dir to test app with separate french msgs
- update handlebars template to include new added french msg
- update test app gulpfile to include extra msg dir
- update test-sample-app.js to scan for added french msg

euwest committed:
extend translators parseMessages fn to search all `msgDirs`
- add `createPoParseFn` to extract parsing out of makeMessages
- change parseMessages to use map/reduce/assign to parse and merge all `po` files

euwest commented:
@beck

kahnvex commented on lib/translator.js:
```diff
@@ -36,19 +38,37 @@ Translator.prototype.getLocales = function() {
+    try{
+      po = fs.readFileSync(fp, {encoding: 'utf8'});
+    }catch(err) {
+      if(err.code !== 'ENOENT') {
```
Why pass over the error if its `ENOENT`?

euwest replied:
otherwise every source directory containing `.po` files(i.e. `msgDir` in the above function) needs to have a `/<language>/LC_MESSAGES/messages.po` or `fs.readFileSync` throws a `ENOENT`. this try/catch allows you to include source directories with translations for only some of your supported languages without blowing up your translation task. maybe we don't care/want that though? I could be convinced to go either way.

beck replied:
setting `po` to empty string results in this function returning `undefined`.  Lets make our functions always return useable results.

Suggesting:

``` js
var hasPo = fs.existsSync(fp);
var po = hasPo ? fs.readFileSync(fp, {encoding: 'utf8'}) : null;
var catalog = po ? gettextParser.po.parse(po).translations[''] : {};
return catalog;
```

beck commented on lib/translator.js:
```diff
@@ -11,9 +11,11 @@ var TranslatedFile = require('./translated-file');
 function Translator(options) {
   this.options = lodash.defaults(options, {
       localeDir: 'locale',
+      msgDirs: ['locale'],
```
Two options doesn't make sense, I would unify on `localeDirs`.

Normally I would be super cautious about altering the options (especially on a public repo) but the plugin is still ~v000 and not yet implemented anywhere so I feel it's appropriate to make significant changes as this thing find's it's stride.

euwest replied:
i saw `localeDir` as a sort of canonical dir of languages, but I guess we can modify the `getLocales` to compile that list by iterating over all the locale directories and filtering for uniqueness.

beck replied:
> I saw localeDir as a sort of canonical dir of languages

I like this idea, it can still be done.  The order of the locale directories will always be important (if a msgid exists in multiple catalogs you want the first encountered to win).  We can do the same with the language list, the first locale dir in the list decides which languages we prefer.

If we end up having a lot of translated packages, we could make configuration super simple:

```
localeDirs: ['locale', 'node_modules/**/locale']
```

beck commented on test/fixtures/app/nested-msgs/locale/fr/LC_MESSAGES/messages.po:
```diff
@@ -0,0 +1,6 @@
+msgid ""
```
A personal preference: I like it when the tests' sample app models real word example.  `nested-msgs` does not mean much to me in terms of "realworldness".  Recommending `installed_deps/foo`

euwest committed:
rename nested-msgs dir to model "realworldness"

euwest committed:
remove msgDirs, modify localeDir to handle msgDirs job
- localeDirs replaces both msgDirs and localeDir
- source first localeDir in localeDirs for canonical list of languages

euwest committed:
change lodash.assign => lodash.defaults for left => right hierarchy
- passing the apps locale directory first in `localeDirs` means its translations trump installables

euwest committed:
add second translation task for testing single and multi catalog

euwest committed:
add message in secondary locale dir to test prioritization

euwest committed:
refactor test.js to add multi-dir tests, reduce repetition

euwest committed:
refactor test-sample-app.js to test single and multi-dir

beck commented on test/fixtures/app/gulpfile.js:
```diff
@@ -15,16 +14,23 @@ gulp.task('build', ['clean'], function(){
 });
 
-gulp.task('default', ['translate']);
+gulp.task('translatesingledir', ['build'], function(){
```
The name changes to the tasks here don't make much sense to me.

What about keeping `translate` unchanged and renaming this one `translate-with-deps`

euwest committed:
rename gulp tasks for test app

beck commented on lib/translator.js:
```diff
@@ -36,19 +37,37 @@ Translator.prototype.getLocales = function() {
+  };
+};
+
+var createMsgParseFn = function(defaultLang, localeDirs) {
```
A few critiques about the name `createMsgParseFn`.

`parseMessages` was a bad choice on my part.  All apologies.  We're not just parsing messages, we're doing more than that, we're parsing entire translation catalog.  I should have used the name `parseCatalog`.

Avoid using the word `Fn` or `Function` when naming items, we'd never want to see `function parseFunction(){....`, we'd expect `function parser()` so lets name this guy `createCatalogParser` (or my personal pref is to use the prefix `get` vs `create`, it is shorter and I generally associate the word `create` with some sort of persistence layer).

Lets avoid anon functions and give these things names:

``` js
function getCatalogParser(defaultLang, localeDirs) {
  var catalogParser = function(lang) {
    ...
  }
  return catalogParser;
}
```

This may be a bit more verbose but what we're doing here is super dense.

Thoughts?

euwest replied:
all that seems reasonable. new names OTW

beck commented on test/test.js:
```diff
@@ -33,20 +32,48 @@ describe('StaticI18n', function(){
+  };
+
+  it('consumes translation catalogs', function() {
+    testGetLocales(singleOptions);
```
Keeping the assertion in a helper function makes it difficult to read what the test is actually doing.

If you look at our tests, you'll often see duplication because we don't want to generalize.  Each test should be their own discrete bit of work.

Rewrite these guys without the helper functions.  It looks like you're trying to test using multiple locales, it may be best to introduce a new describe block:

```
describe('Translator', function(){
  describe('configured with multiple locale dirs', function() {
    it('merges catalogs that belong to the same language', function() {
      ...
```

If you need help with this, let me know.

euwest committed:
rename parser function constructors

euwest committed:
add describe blocks for single/multi locale directory cases

euwest committed:
replace try block with fs.existSync

euwest commented:
:bell:

beck commented:
:+1:
